ddcommon-ffi/src/slice_mut.rs

```diff
@@ -0,0 +1,295 @@
+        unsafe { MutSlice::from_raw_parts(self.ptr.cast(), self.len) }.as_mut_slice()
+    }
+}
+
+impl<'a, T: 'a> MutSlice<'a, T> {
+    /// Creates a valid empty slice (len=0, ptr is non-null).
+    #[must_use]
+    // TODO: this can't be const under 1.78, but can under later versions.
+    pub fn empty() -> Self {
+        Self {
+            ptr: [].as_mut_ptr(),
+            len: 0,
+            _marker: PhantomData,
+        }
+    }
+
+    /// # Safety
+    /// Uphold the same safety requirements as [std::str::from_raw_parts].
+    /// However, it is allowed but not recommended to provide a null pointer
+    /// when the len is 0.
+    pub unsafe fn from_raw_parts(ptr: *mut T, len: usize) -> Self {
+        Self {
+            ptr,
+            len,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn new(slice: &mut [T]) -> Self {
+        Self {
+            ptr: slice.as_mut_ptr(),
+            len: slice.len(),
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn as_mut_slice(&mut self) -> &'a mut [T] {
+        if !self.ptr.is_null() {
+            // Crashing immediately is likely better than ignoring these.
+            assert!(is_aligned(self.ptr));
+            assert!(self.len <= isize::MAX as usize);
+            unsafe { slice::from_raw_parts_mut(self.ptr, self.len) }
+        } else {
+            // Crashing immediately is likely better than ignoring this.
+            assert_eq!(self.len, 0);
+            &mut []
+        }
+    }
+
+    pub fn as_slice(&self) -> &'a [T] {
+        if !self.ptr.is_null() {
+            // Crashing immediately is likely better than ignoring these.
+            assert!(is_aligned(self.ptr));
+            assert!(self.len <= isize::MAX as usize);
+            unsafe { slice::from_raw_parts(self.ptr, self.len) }
+        } else {
+            // Crashing immediately is likely better than ignoring this.
+            assert_eq!(self.len, 0);
+            &[]
+        }
+    }
+
+    pub fn into_slice(self) -> &'a [T] {
+        self.as_slice()
+    }
+
+    pub fn into_mut_slice(mut self) -> &'a mut [T] {
+        self.as_mut_slice()
+    }
+}
+
+impl<T> Default for MutSlice<'_, T> {
+    fn default() -> Self {
+        Self::empty()
+    }
+}
+
+impl<'a, T> Hash for MutSlice<'a, T>
+where
+    MutSlice<'a, T>: AsBytes<'a>,
+{
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write(self.as_bytes())
+    }
+}
+
+impl<'a, T> serde::Serialize for MutSlice<'a, T>
+where
+    MutSlice<'a, T>: AsBytes<'a>,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self.try_to_utf8().map_err(Error::custom)?)
+    }
+}
+
+impl<'a, T> Display for MutSlice<'a, T>
+where
+    MutSlice<'a, T>: AsBytes<'a>,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.try_to_utf8().map_err(|_| std::fmt::Error)?)
+    }
+}
```

Comment:
What do we need Display for on a MutSlice?

Reply:
We don't.  I just copied over all the functions from `Slice`.  I can remove it in the YAGNI philosophy